wolfcrypt/src/ecc.c

```diff
@@ -5756,6 +5821,264 @@ int wc_ecc_sign_hash_ex(const byte* in, word32 inlen, WC_RNG* rng,
+    /* bail out if any error has been hit at this point */
+    if (ret != 0) {
+        if (x != NULL)
+            XFREE(x, heap, DYNAMIC_TYPE_PRIVATE_KEY);
+        if (K != NULL)
+            XFREE(K, heap, DYNAMIC_TYPE_ECC_BUFFER);
+        if (V != NULL)
+            XFREE(V, heap, DYNAMIC_TYPE_ECC_BUFFER);
+        if (h1 != NULL)
+            XFREE(h1, heap, DYNAMIC_TYPE_DIGEST);
+        return ret;
+    }
+#endif
+
+    VSz = KSz = h1len = hashSz;
+    XMEMSET(V, 0x01, VSz);
+    XMEMSET(K, 0x00, KSz);
+
+    ret = mp_to_unsigned_bin(priv, x);
+
+    mp_init(&z1); /* always init z1 and free z1 */
+    if (ret == 0) {
+        qbits = mp_count_bits(order);
+    }
+
+    if (ret == 0) {
+        ret = mp_read_unsigned_bin(&z1, hash, hashSz);
+    }
+
+    if (ret == 0) {
+        /* right shift by bits in hash minus bits in order */
+        mp_rshb(&z1, (hashSz * WOLFSSL_BIT_SIZE) - qbits);
+        XMEMSET(h1, 0, sizeof(h1));
+    }
+
+    if (ret == 0) {
+        /* mod reduce by order using conditional subtract */
+        if (mp_cmp(&z1, order) == MP_GT) {
+            mp_sub(&z1, order, &z1);
+            h1len = mp_unsigned_bin_size(&z1);
+            if (h1len < 0 || h1len > WC_MAX_DIGEST_SIZE) {
+                ret = BUFFER_E;
+            }
+            else {
+                ret = mp_to_unsigned_bin(&z1, h1);
+            }
+        }
+        else {
+            /* use original hash and keep leading 0's */
+            h1len = hashSz;
+            XMEMCPY(h1, hash, hashSz);
+        }
+    }
+    mp_free(&z1);
+
+    /* step d. */
+    if (ret == 0) {
+        intOct = 0x00;
+        ret = _HMAC_K(K, KSz, V, VSz, h1, h1len, x, xSz, &intOct, K,
+                hashType, heap);
+    }
+
+    /* step e. */
+    if (ret == 0) {
+        ret = _HMAC_K(K, KSz, V, VSz, NULL, 0, NULL, 0, NULL, V, hashType,
+                heap);
+    }
+
+
+    /* step f. */
+    if (ret == 0) {
+        intOct = 0x01;
+        ret = _HMAC_K(K, KSz, V, VSz, h1, h1len, x, xSz, &intOct, K, hashType,
+                heap);
+    }
+
+    /* step g. */
+    if (ret == 0) {
+        ret = _HMAC_K(K, KSz, V, VSz, NULL, 0, NULL, 0, NULL, V, hashType,
+                heap);
+    }
+
+    /* step h. */
+    if (ret == 0 ) {
+        int err;
+        intOct = 0x00;
+        do {
+            err = 0; /* start as good until generated k is tested */
+
+            ret = _HMAC_K(K, KSz, V, VSz, NULL, 0, NULL, 0, NULL, V, hashType,
+                    heap);
+            if (ret == 0) {
+                mp_clear(k);
+                ret = mp_read_unsigned_bin(k, V, VSz);
+            }
+
+            if ((ret == 0) && ((int)(VSz * WOLFSSL_BIT_SIZE) != qbits)) {
+                /* handle odd case where shift of 'k' is needed */
+                mp_rshb(k, (VSz * WOLFSSL_BIT_SIZE) - qbits);
+            }
+
+            /* the key should be smaller than the order of base point */
+            if (ret == 0) {
+                if (mp_cmp(k, order) != MP_LT) {
```

Comment:
pull up to previous line

Reply:
On this one am not sure about adding it to the previous statement block. The if check above it also has a check for the size of V value which is not needed with the compare.